src/lang/std/artifactGraph.ts

```diff
@@ -0,0 +1,682 @@
+import { PathToNode, Program, SourceRange } from 'lang/wasm'
+import { Models } from '@kittycad/lib'
+import { getNodePathFromSourceRange } from 'lang/queryAst'
+import { err } from 'lib/trap'
+
+interface CommonCommandProperties {
+  range: SourceRange
+  pathToNode: PathToNode
+}
+
+export interface PlaneArtifact {
+  type: 'plane'
+  pathIds: Array<string>
+  codeRef: CommonCommandProperties
+}
+export interface PlaneArtifactRich {
+  type: 'plane'
+  paths: Array<PathArtifact>
+  codeRef: CommonCommandProperties
+}
+
+export interface PathArtifact {
+  type: 'path'
+  planeId: string
+  segIds: Array<string>
+  extrusionId: string
+  solid2dId?: string
+  codeRef: CommonCommandProperties
+}
+
+interface solid2D {
+  type: 'solid2D'
+  pathId: string
+}
+export interface PathArtifactRich {
+  type: 'path'
+  plane: PlaneArtifact | WallArtifact
+  segments: Array<SegmentArtifact>
+  extrusion: ExtrusionArtifact
+  codeRef: CommonCommandProperties
+}
+
+interface SegmentArtifact {
+  type: 'segment'
+  pathId: string
+  surfaceId: string
+  edgeIds: Array<string>
+  edgeCutId?: string
+  codeRef: CommonCommandProperties
+}
+interface SegmentArtifactRich {
+  type: 'segment'
+  path: PathArtifact
+  surf: WallArtifact
+  edges: Array<ExtrudeEdge>
+  edgeCut?: EdgeCut
+  codeRef: CommonCommandProperties
+}
+
+interface ExtrusionArtifact {
+  type: 'extrusion'
+  pathId: string
+  surfaceIds: Array<string>
+  edgeIds: Array<string>
+  codeRef: CommonCommandProperties
+}
+interface ExtrusionArtifactRich {
+  type: 'extrusion'
+  path: PathArtifact
+  surfaces: Array<WallArtifact | CapArtifact>
+  edges: Array<ExtrudeEdge>
+  codeRef: CommonCommandProperties
+}
+
+interface WallArtifact {
+  type: 'wall'
+  segId: string
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+interface CapArtifact {
+  type: 'cap'
+  subType: 'start' | 'end'
+  edgeCutEdgeIds: Array<string>
+  extrusionId: string
+  pathIds: Array<string>
+}
+
+interface ExtrudeEdge {
+  type: 'extrudeEdge'
+  segId: string
+  extrusionId: string
+  edgeId: string
+}
+
+/** A edgeCut is a more generic term for both fillet or chamfer */
+interface EdgeCut {
+  type: 'edgeCut'
+  subType: 'fillet' | 'chamfer'
+  consumedEdgeId: string
+  edgeIds: Array<string>
+  surfaceId: string
+  codeRef: CommonCommandProperties
+}
+
+interface EdgeCutEdge {
+  type: 'edgeCutEdge'
+  edgeCutId: string
+  surfaceId: string
+}
+
+export type Artifact =
+  | PlaneArtifact
+  | PathArtifact
+  | SegmentArtifact
+  | ExtrusionArtifact
+  | WallArtifact
+  | CapArtifact
+  | ExtrudeEdge
+  | EdgeCut
+  | EdgeCutEdge
+  | solid2D
+
+export type ArtifactGraph = Map<string, Artifact>
```

Comment:
In my new branch, I'm thinking about adding this to help communicate what should be an ID into this map.

```suggestion
export type ArtifactId = string

export type ArtifactGraph = Map<ArtifactId, Artifact>
```

Reply:
sounds good

Reply:
Was going to say that `commandId` might be better, but that's only true for plan, path, segment, but not extrude walls caps, other edges. I think I used `commandId` for naming in a few places where I should probably do a rename to `artifactId`